proposal/proposal.tex

```diff
@@ -0,0 +1,136 @@
+\section{The project}
+
+\subsection{Project core}
+First, a CPU-based library for performing individual matrix computations in the tropical
+semiring should be implemented, using a dense representation of matrices.
+Techniques such as vectorisation and tiling should be considered for performance gains.
+
+Matrix multiplication expressions will be generated from a program
+execution's dependency structure as a graph, with one node per register per dynamic instruction,
+organised as a column of nodes per dynamic instruction.
+
+The aforementioned matrix computation library will be used to compute the results
+of these expressions, using techniques such as caching and evaluation order optimisation
+to maximise efficiency.
+
+Finally, a benchmark suite for testing performance of implementations will be implemented,
+and a baseline performance will be established using a \lq{classical\rq} technique for computing
+dependency chains (i.e. simulating one instruction at a time naively, computing its dependencies
+one-by-one).
+
+\subsection{Project extensions}
+
+\subsubsection{Sparse matrix representation}
+The core of the project implements a matrix computation library using a dense
+representation of matrices. In an effort to improve memory usage, I will implement a
+similar CPU-based matrix computation library using a sparse representation of matrices
+as an extension to the project core.
+
+The sparse library will be benchmarked against the dense library for register-only
+program execution dependency graphs.
+
+\subsubsection{GPU compute}
+The core of the project implements a CPU-based matrix computation library.
+To exploit the excellent parallel power of GPUs, I will implement two GPU-based matrix
+computation libraries, for dense and sparse representations of matrices, using a
+programming model such as OpenCL or CUDA.
+
+The GPU libraries will be benchmarked against their CPU counterparts for register-only
+program execution dependency graphs.
+
+\subsubsection{Through-memory dependencies}
+So far we have only computed results of matrix expressions generated from register-only
+program execution dependency graphs. However, for real programs, the true critical path
+often involves memory operations.
+
+A simple approach is to naively include memory locations as graph nodes alongside
+registers for each dynamic instruction, but this results in a catastrophic explosion of
+matrix sizes. Nevertheless, these matrices are extremely sparse, as most memory operations
+access only two or three locations at a time, and so are usable with sparse representations.
+For this approach, I will benchmark the performance of the two previously implemented sparse
+matrix computation libraries (running on CPU and GPU) against the baseline.
+
+An alternative approach is to model the effects of memory operations using tropical semiring
+\textit{addition} rather than just multiplication, as described in \href{https://www.google.com}{Section II-B of Chadwick (2025)}.
```

Comment:
I've now put my little paper at https://www.cl.cam.ac.uk/~awc32/tropical-ilp.pdf :blush:
Your citations should be done with bibtex or similar, so inline you have a [1] or [Chadwick, 2025] or whatever, and then at the end you have the full list.